compact.go

```diff
@@ -682,6 +692,61 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta,
 	return nil
 }
 
+func (c *LeveledCompactor) mergeChunks(chks []chunks.Meta, numSamples int) []chunks.Meta {
+	newChks := make([]chunks.Meta, 0, len(chks))
+	for i := 0; i < len(chks); i++ {
+		if i >= len(chks)-1 || chks[i].Chunk.NumSamples()+chks[i+1].Chunk.NumSamples() > numSamples {
+			newChks = append(newChks, chks[i])
+			continue
+		}
+		newChunk := chunkenc.NewXORChunk()
+		app, err := newChunk.Appender()
+		if err != nil {
+			level.Warn(c.logger).Log("msg", "failed to merge chunks", "err", err)
+			return chks
+		}
+		if err := appendChunk(app, chks[i].Chunk); err != nil {
+			level.Warn(c.logger).Log("msg", "failed to merge chunks", "err", err)
+			return chks
+		}
+		minTime := chks[i].MinTime
+		maxTime := chks[i].MaxTime
+
+		for i < len(chks)-1 && newChunk.NumSamples()+chks[i+1].Chunk.NumSamples() <= numSamples {
```

Comment:
can you explain what is happening here and will try to simplify.

Reply:
Added a comment, and did some refactoring. Basically it is trying to merge as many chunks as possible. If there are 4 120 sample chunks, and a goal of 360 samples per chunk it will merge the next two chunks (i=1, i=2) with the chunk at i=0.

Happy to simplify further!